demos/coreHTTP/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,965 @@
+     * can be sent over the same established TCP connection. This is done in
+     * order to download the file in parts. */
+    xRequestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Get the length of the S3 file. */
+    xStatus = prvGetS3ObjectFileSize( &xRequestInfo,
+                                      xRequestQueue,
+                                      xResponseQueue,
+                                      &xFileSize );
+
+    if( xFileSize < democonfigRANGE_REQUEST_LENGTH )
+    {
+        xNumReqBytes = xFileSize;
+    }
+    else
+    {
+        xNumReqBytes = democonfigRANGE_REQUEST_LENGTH;
+    }
+
+    /* Here we iterate sending byte range requests and retrieving responses
+     * until the full file has been downloaded. We keep track of the next byte
+     * to download with xCurByte. When this reaches the xFileSize we stop
+     * downloading. We keep track of the number of responses we are waiting for
+     * with xRequestCount.
+     */
+    while( ( xStatus != pdFAIL ) && ( xCurByte < xFileSize || xRequestCount > 0 ) )
+    {
+        /* Send range request if remaining. */
+        if( xCurByte < xFileSize )
+        {
+            xStatus = prvRequestS3ObjectRange( &xRequestInfo,
+                                               xRequestQueue,
+                                               xCurByte,
+                                               xCurByte + xNumReqBytes - 1 );
+
+            if( xStatus == pdPASS )
+            {
+                xRequestCount += 1;
+                xCurByte += xNumReqBytes;
+
+                if( ( xFileSize - xCurByte ) < xNumReqBytes )
+                {
+                    xNumReqBytes = xFileSize - xCurByte;
+                }
+            }
+        }
+
+        /* Retrieve response. */
+        if( ( xRequestCount > 0 ) && ( xStatus != pdFAIL ) )
+        {
+            xStatus = prvRetrieveHTTPResponse( xResponseQueue, &xResponseItem );
+
+            if( xStatus == pdPASS )
+            {
+                LogInfo( ( "Main thread received HTTP response" ) );
+                LogDebug( ( "Response Headers:\n%.*s",
+                            ( int32_t ) xResponseItem.xResponse.headersLen,
+                            xResponseItem.xResponse.pHeaders ) );
+                LogDebug( ( "Response Status:\n%u", xResponseItem.xResponse.statusCode ) );
+                LogDebug( ( "Response Body:\n%.*s\n", ( int32_t ) xResponseItem.xResponse.bodyLen,
+                            xResponseItem.xResponse.pBody ) );
+
+                if( xResponseItem.xResponse.statusCode != httpexampleHTTP_STATUS_CODE_PARTIAL_CONTENT )
+                {
+                    LogError( ( "Received response with unexpected status code: %d", xResponseItem.xResponse.statusCode ) );
+                    xStatus = pdFAIL;
+                }
+                else
+                {
+                    xRequestCount -= 1;
+                }
+            }
```

Comment:
More comments in this function would be better to improve the readability, especially in the while loop.